src/DataStructures/StaticDeque.hpp

```diff
@@ -0,0 +1,520 @@
+StaticDeque<T, Capacity>::~StaticDeque() {
+  clear();
+}
+
+template <typename T, size_t Capacity>
+auto StaticDeque<T, Capacity>::operator=(const StaticDeque& other)
+    -> StaticDeque& {
+  if (size() > other.size()) {
+    resize(other.size());
+  }
+  size_t i = 0;
+  for (; i < size(); ++i) {
+    (*this)[i] = other[i];
+  }
+  for (; i < other.size(); ++i) {
+    push_back(other[i]);
+  }
+  return *this;
+}
+
+template <typename T, size_t Capacity>
+auto StaticDeque<T, Capacity>::operator=(StaticDeque&& other) -> StaticDeque& {
+  if (size() > other.size()) {
+    resize(other.size());
+  }
+  size_t i = 0;
+  for (; i < size(); ++i) {
+    (*this)[i] = std::move(other[i]);
+  }
+  for (; i < other.size(); ++i) {
+    push_back(std::move(other[i]));
+  }
+  return *this;
+}
+
+template <typename T, size_t Capacity>
+auto StaticDeque<T, Capacity>::operator=(const std::initializer_list<T> init)
+    -> StaticDeque& {
+  assign(init);
+  return *this;
+}
+
+template <typename T, size_t Capacity>
+template <typename InputIterator,
+          Requires<static_deque_detail::is_input_iterator<InputIterator>>>
+void StaticDeque<T, Capacity>::assign(InputIterator first,
+                                      const InputIterator last) {
+  // This could be slightly more efficient if this object is not
+  // already empty by assigning to existing objects.
+  clear();
+  while (first != last) {
+    push_back(*first++);
+  }
+}
+
+template <typename T, size_t Capacity>
+void StaticDeque<T, Capacity>::assign(const size_type count, const T& value) {
+  ASSERT(count <= max_size(), count << " exceeds max size " << max_size());
+  // This could be slightly more efficient if this object is not
+  // already empty by assigning to existing objects.
+  clear();
+  resize(count, value);
+}
+
+template <typename T, size_t Capacity>
+void StaticDeque<T, Capacity>::assign(const std::initializer_list<T> init) {
+  assign(init.begin(), init.end());
+}
+
+template <typename T, size_t Capacity>
+void StaticDeque<T, Capacity>::resize(const size_type count) {
+  ASSERT(count <= max_size(), count << " exceeds max size " << max_size());
+  while (size() > count) {
+    pop_back();
+  }
+  while (size() < count) {
+    emplace_back();
+  }
+}
+
+template <typename T, size_t Capacity>
+void StaticDeque<T, Capacity>::resize(const size_type count, const T& value) {
+  ASSERT(count <= max_size(), count << " exceeds max size " << max_size());
+  while (size() > count) {
+    pop_back();
+  }
+  while (size() < count) {
+    push_back(value);
+  }
+}
+
+template <typename T, size_t Capacity>
+template <typename... Args>
+T& StaticDeque<T, Capacity>::emplace_front(Args&&... args) {
+  ASSERT(size() < max_size(), "Container is full.");
+  auto* const new_entry =
+      new (entry_address(Capacity - 1)) T(std::forward<Args>(args)...);
+  // This must be done last in case T's constructor throws an
+  // exception.
+  if (start_position_ == 0) {
+    start_position_ = Capacity - 1;
```

Comment:
Another thing to make sure I understand, and that might be useful to state in a C++ comment (sorry if I missed the comment!): here rather than touching every element, you use a `start_position_` to keep track of where in the static buffer you start from, then pushing to the front can wrap to the end of the buffer, but pushing to the back will never wrap? Maybe add to the `data_` comment: `This means that for adding data to the front we need to check if our start position is 0, since then we start adding at the end and update start_position_ to the end of the byte stream.` ?